### ISSUE TITLE
How to predict multi-dimensional time series

### ISSUE THREAD

**Issue by valer1435** [valer1435/RepoPilot#11]
How to forecast time series with fedot? @RepoPilotAssistant

**Comment by RepoPilotAssistant** [comment]
It seems like the chat history provided does not contain any specific question or context to derive an answer from. Please provide more details or the specific question you would like answered regarding the Fedot library or the GitHub issue.

**Comment by valer1435** [comment]
How to forecast multi-dimensional time series? @RepoPilotAssistant

**Comment by RepoPilotAssistant** [comment]
To forecast a multi-dimensional time series using Fedot, you can follow the example provided in the context for multivariate time-series forecasting. Here is a concise example:

```python
from fedot.api.main import Fedot
from fedot.core.data.data import InputData
from fedot.core.data.data_split import train_test_data_setup
from fedot.core.repository.tasks import Task, TaskTypesEnum, TsForecastingParams

# Name of the column with target time-series
target = 'col_3'

# Set task type and forecast length
task = Task(TaskTypesEnum.ts_forecasting,
            TsForecastingParams(forecast_length=20))  # Forecast_length - required depth of forecast

# Load data from csv
data = InputData.from_csv_multi_time_series(
    file_path='time_series.csv',
    task=task,
    target_column=target,
    columns_to_use=['col_1', 'col_2', 'col_3', ..., 'col_n'])

# Split data for in-sample forecast
train_data, test_data = train_test_data_setup(data, validation_blocks=2)

# Init model for the time-series forecasting
model = Fedot(problem='ts_forecasting',
              task_params=task.task_params,
              timeout=5,
              n_jobs=-1,
              cv_folds=2)

# Run AutoML model design
pipeline = model.fit(train_data)
# Plot obtained pipeline
pipeline.show()

# Use model to obtain two-step in-sample forecast
forecast = model.predict(test_data, validation_blocks=2)
print(model.get_metrics(metric_names=['rmse', 'mae', 'mape'],
                        target=test_data.target, validation_blocks=2))
```

This example demonstrates how to use multiple time-series to forecast one target time-series. The `InputData.from_csv_multi_time_series` method is used to load the data, and the `Fedot` class is initialized with the appropriate parameters for time-series forecasting. The pipeline is then fitted on the training data, and predictions are made on the test data. Metrics are calculated to evaluate the model's performance.

**Comment by RepoPilotAssistant** [comment]
To forecast multi-dimensional time series using FEDOT, you can use the multi-modal approach. This approach involves creating a data source node for each time series. Here is an example code snippet demonstrating how to do this:

```python
import numpy as np
import pandas as pd

from fedot.api.main import Fedot
from fedot.core.repository.tasks import TsForecastingParams

forecast_length = 10

# load data from csv
df_train = pd.read_csv('data_train.csv')

# get time-series
ws_history = np.ravel(np.array(df_train['wind_speed']))[:200]
ssh_history = np.ravel(np.array(df_train['sea_height']))[:200]

# use dict as input data
historical_data = {
    'ws': ws_history,  # additional variable
    'ssh': ssh_history,  # target variable
}

# init model for the time-series forecasting
fedot = Fedot(problem='ts_forecasting',
              task_params=TsForecastingParams(forecast_length=forecast_length),
              timeout=10)

# run AutoML model design
pipeline = fedot.fit(features=historical_data,
                     target=ssh_history)  # specify target time-series

# get in-sample forecast
fedot.predict(historical_data)

# get metrics
metric = fedot.get_metrics(ssh_history[-forecast_length:])

pipeline.show()
fedot.plot_prediction(target='ssh')
```

In this example:
- `ws_history` and `ssh_history` are the historical time series data.
- `historical_data` is a dictionary where each key-value pair represents a time series.
- The `Fedot` class is initialized with the problem type `ts_forecasting` and the forecast length.
- The `fit` method is used to train the model on the historical data.
- The `predict` method is used to obtain the forecast.
- The `get_metrics` method calculates the performance metrics.
- The `pipeline.show()` method displays the pipeline structure.
- The `fedot.plot_prediction(target='ssh')` method plots the forecast.

**Comment by valer1435** [comment]
Please explain more detailed what the difference between insample and outsample forecasting?
 @RepoPilotAssistant

**Comment by RepoPilotAssistant** [comment]
In-sample forecasting refers to making predictions within the same dataset that was used for training the model. This means that the model uses the historical data it was trained on to generate forecasts for the same time period or a subset of it. In-sample forecasting is useful for validating the model's performance on known data but does not provide a true test of its predictive capabilities on unseen data.

Out-of-sample forecasting, on the other hand, involves making predictions on data that was not used during the training phase. This is a more realistic scenario as it tests the model's ability to generalize to new, unseen data. Out-of-sample forecasting is crucial for assessing the real-world applicability of the model and its predictive accuracy.

In the provided context, the example demonstrates in-sample forecasting where the model is used to predict values within the same historical data it was trained on. For out-of-sample forecasting, you would need to use a separate test dataset that was not part of the training process. Here is an example of how you might modify the code for out-of-sample forecasting:

```python
# Split data into train and test sets
train_size = int(len(ssh_history) * 0.8)
train, test = ssh_history[:train_size], ssh_history[train_size:]

# Use only the train set for training
historical_data_train = {
    'ws': ws_history[:train_size],
    'ssh': train,
}

# Init model for the time-series forecasting
fedot = Fedot(problem='ts_forecasting',
              task_params=TsForecastingParams(forecast_length=forecast_length),
              timeout=10)

# Run AutoML model design
pipeline = fedot.fit(features=historical_data_train,
                     target=train)  # specify target time-series

# Prepare test data for prediction
historical_data_test = {
    'ws': ws_history[train_size:],
    'ssh': test,
}

# Get out-of-sample forecast
forecast = fedot.predict(historical_data_test)

# Get metrics
metric = fedot.get_metrics(test[-forecast_length:])

pipeline.show()
fedot.plot_prediction(target='ssh')
```

In this modified example:
- The data is split into training and test sets.
- The model is trained on the training set.
- The `predict` method is used to obtain forecasts on the test set, which is out-of-sample data.
- The `get_metrics` method calculates the performance metrics on the test set.
- The `pipeline.show()` method displays the pipeline structure.
- The `fedot.plot_prediction(target='ssh')` method plots the forecast.

**Comment by valer1435** [comment]
Can you provide explanation of predict method? @RepoPilotAssistant

**Comment by RepoPilotAssistant** [comment]
The `predict` method in FEDOT is used to generate forecasts based on the trained model. Here is a detailed explanation of the `predict` method and its parameters:

### `predict` Method

```python
fedot.predict(historical_data, validation_blocks=None)
```

#### Parameters:

1. **`historical_data`**:
   - **Type**: `dict` or `InputData`
   - **Description**: The historical data used for generating forecasts. This can be a dictionary where each key-value pair represents a time series, or an `InputData` object.

2. **`validation_blocks`** (optional):
   - **Type**: `int`
   - **Description**: The number of validation blocks to use for multi-step forecasting. This parameter is useful for out-of-sample forecasting when you want to generate forecasts for multiple steps ahead.

#### Example Usage:

```python
# Get in-sample forecast
fedot.predict(historical_data)

# Get out-of-sample forecast with validation blocks
fedot.predict(historical_data_test, validation_blocks=2)
```

### Detailed Explanation:

- **In-sample Forecast**: When you call `fedot.predict(historical_data)`, the model uses the historical data it was trained on to generate forecasts. This is typically used for validating the model's performance on known data.

- **Out-of-sample Forecast**: When you provide a separate test dataset and specify `validation_blocks`, the model generates forecasts for multiple steps ahead. This is useful for assessing the model's ability to generalize to new, unseen data.

### Example Code:

Here is an example demonstrating both in-sample and out-of-sample forecasting:

```python
import numpy as np
import pandas as pd

from fedot.api.main import Fedot
from fedot.core.repository.tasks import TsForecastingParams

forecast_length = 10

# load data from csv
df_train = pd.read_csv('data_train.csv')

# get time-series
ws_history = np.ravel(np.array(df_train['wind_speed']))[:200]
ssh_history = np.ravel(np.array(df_train['sea_height']))[:200]

# use dict as input data
historical_data = {
    'ws': ws_history,  # additional variable
    'ssh': ssh_history,  # target variable
}

# init model for the time-series forecasting
fedot = Fedot(problem='ts_forecasting',
              task_params=TsForecastingParams(forecast_length=forecast_length),
              timeout=10)

# run AutoML model design
pipeline = fedot.fit(features=historical_data,
                     target=ssh_history)  # specify target time-series

# get in-sample forecast
fedot.predict(historical_data)

# get metrics
metric = fedot.get_metrics(ssh_history[-forecast_length:])

# Split data into train and test sets
train_size = int(len(ssh_history) * 0.8)
train, test = ssh_history[:train_size], ssh_history[train_size:]

# Use only the train set for training
historical_data_train = {
    'ws': ws_history[:train_size],
    'ssh': train,
}

# Init model for the time-series forecasting
fedot = Fedot(problem='ts_forecasting',
              task_params=TsForecastingParams(forecast_length=forecast_length),
              timeout=10)

# Run AutoML model design
pipeline = fedot.fit(features=historical_data_train,
                     target=train)  # specify target time-series

# Prepare test data for prediction
historical_data_test = {
    'ws': ws_history[train_size:],
    'ssh': test,
}

# Get out-of-sample forecast with validation blocks
forecast = fedot.predict(historical_data_test, validation_blocks=2)

# Get metrics
metric = fedot.get_metrics(test[-forecast_length:])

pipeline.show()
fedot.plot_prediction(target='ssh')
```

In this example:
- The `predict` method is used to generate both in-sample and out-of-sample forecasts.
- For in-sample forecasting, the model uses the historical data it was trained on.
- For out-of-sample forecasting, the model uses a separate test dataset and generates forecasts for multiple steps ahead using `validation_blocks`.

**Comment by RepoPilotAssistant** [comment]
To forecast time series using Fedot, you can use the automated mode to get a pipeline with automatically composed architecture and tuned hyperparameters. Here is a concise example:

```python
from fedot.api.main import Fedot
from fedot.core.data.data import InputData
from fedot.core.data.data_split import train_test_data_setup
from fedot.core.repository.tasks import Task, TaskTypesEnum, TsForecastingParams

# Specify the task and the forecast length (required depth of forecast)
task = Task(TaskTypesEnum.ts_forecasting, TsForecastingParams(forecast_length=10))

# Load data from csv
train_input = InputData.from_csv_time_series(task=task, file_path='time_series.csv', delimiter=',', target_column='value')

# Split data for train and test
train_data, test_data = train_test_data_setup(train_input)

# Init model for the time-series forecasting
model = Fedot(problem='ts_forecasting', task_params=task.task_params)

# Run AutoML model design
pipeline = model.fit(train_data)

# Plot obtained pipeline
pipeline.show()

# Use model to obtain out-of-sample forecast with one step
forecast = model.forecast(test_data)

# Print metrics
print(model.get_metrics(metric_names=['rmse', 'mae', 'mape'], target=test_data.target))

# Plot forecasting result
model.plot_prediction()
```

This example demonstrates how to:
1. Define the forecasting task and specify the forecast length.
2. Load time series data from a CSV file.
3. Split the data into training and testing sets.
4. Initialize the Fedot model for time series forecasting.
5. Fit the model to the training data to automatically design the pipeline.
6. Obtain and plot the out-of-sample forecast.
7. Evaluate the model's performance using metrics like RMSE, MAE, and MAPE.
8. Plot the forecasting result.

**Comment by valer1435** [comment]
Please give explanation what the difference between insample and out of sample forecasting

**Comment by valer1435** [comment]
Please give explanation what the difference between insample and out of sample forecasting @RepoPilotAssistant

**Comment by RepoPilotAssistant** [comment]
In-sample forecasting involves predicting known parts of the time series using the trained model, where the model forecasts values within the training sample. This approach uses historical data up to a certain point to predict the subsequent known values. Out-of-sample forecasting, on the other hand, predicts future values that are not part of the training data, using the entire training sample as prehistory. The model iteratively predicts values, using previously forecasted values to predict the next ones when the forecast horizon exceeds the forecast length the model was trained for.

**Comment by valer1435** [comment]
So if I want to make out of sample forecast in fedot what should I do?

**Comment by valer1435** [comment]
So if I want to make out of sample forecast in fedot what should I do? @RepoPilotAssistant

**Comment by RepoPilotAssistant** [comment]
To make an out-of-sample forecast in Fedot, you need to follow these steps:

1. Set the task type and forecast length using `Task` and `TsForecastingParams`.
2. Load the data from a CSV file using `InputData.from_csv_time_series`.
3. Split the data for out-of-sample forecast using `train_test_data_setup`.
4. Initialize the model for time-series forecasting with `Fedot`.
5. Run AutoML model design by fitting the model on the training data.
6. Use the model to obtain the out-of-sample forecast with the desired horizon using the `forecast` method.
7. Optionally, get metrics for the prediction using `get_metrics` and plot the forecasting result with `plot_prediction`.

Here is a simplified example:

```python
from fedot.api.main import Fedot
from fedot.core.data.data import InputData
from fedot.core.data.data_split import train_test_data_setup
from fedot.core.repository.tasks import Task, TaskTypesEnum, TsForecastingParams

# Set task type and forecast length
task = Task(TaskTypesEnum.ts_forecasting, TsForecastingParams(forecast_length=10))

# Load data from CSV
train_input = InputData.from_csv_time_series(file_path='time_series.csv', task=task, target_column='value')

# Split data for out-of-sample forecast
train_data, test_data = train_test_data_setup(train_input)

# Initialize model for time-series forecasting
model = Fedot(problem='ts_forecasting', task_params=task.task_params, cv_folds=2)

# Run AutoML model design
pipeline = model.fit(train_data)

# Obtain out-of-sample forecast
out_of_sample_forecast = model.forecast(test_data)

# Get metrics for the prediction
print('Metrics for out-of-sample forecast: ', model.get_metrics(metric_names=['rmse', 'mae', 'mape']))

# Plot forecasting result
model.plot_prediction()
```

This example demonstrates how to perform an out-of-sample forecast with the default horizon. For custom horizons, you can specify the `horizon` parameter in the `forecast` method.

**Comment by RepoPilotAssistant** [comment]
To make an out-of-sample forecast in Fedot, you need to follow these steps:

1. Set the task type and forecast length using `Task` and `TsForecastingParams`.
2. Load the time series data from a CSV file using `InputData.from_csv_time_series`.
3. Split the data into training and testing sets using `train_test_data_setup`.
4. Initialize the Fedot model for time-series forecasting with the specified problem type and task parameters.
5. Run the AutoML model design by fitting the model to the training data.
6. Obtain the out-of-sample forecast using the `forecast` method on the test data.
7. Optionally, get metrics for the forecast using `get_metrics` and plot the forecasting result using `plot_prediction`.

Here is a simplified example:

```python
from fedot.api.main import Fedot
from fedot.core.data.data import InputData
from fedot.core.data.data_split import train_test_data_setup
from fedot.core.repository.tasks import Task, TaskTypesEnum, TsForecastingParams

# Set task type and forecast length
task = Task(TaskTypesEnum.ts_forecasting, TsForecastingParams(forecast_length=10))

# Load data from CSV
train_input = InputData.from_csv_time_series(file_path='time_series.csv', task=task, target_column='value')

# Split data for out-of-sample forecast
train_data, test_data = train_test_data_setup(train_input)

# Initialize model for time-series forecasting
model = Fedot(problem='ts_forecasting', task_params=task.task_params, cv_folds=2)

# Run AutoML model design
pipeline = model.fit(train_data)

# Obtain out-of-sample forecast
out_of_sample_forecast = model.forecast(test_data)

# Get metrics for the prediction
print('Metrics for out-of-sample forecast: ', model.get_metrics(metric_names=['rmse', 'mae', 'mape']))

# Plot forecasting result
model.plot_prediction()
```
